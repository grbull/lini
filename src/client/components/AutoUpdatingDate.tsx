import React, { ReactElement, useState } from 'react';

import { useInterval } from '../hooks/useInterval';

interface Props {
  className?: string;
  date: string | null;
  format: (date: string) => string;
}

export function AutoUpdatingDate({
  className,
  date,
  format,
}: Props): ReactElement {
  const [formattedDate, setFormattedDate] = useState(
    date ? format(date) : 'N/A'
  );

  useInterval(() => {
    if (date) {
      setFormattedDate(format(date));
    }
    // Updates date every 10min.
  }, 1000 * 60 * 10);

  return (
    <time className={className} dateTime={date || undefined}>
      {formattedDate}
    </time>
  );
}

AutoUpdatingDate.defaultProps = { className: undefined };
