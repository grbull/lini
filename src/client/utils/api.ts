import Axios from 'axios';

import { AuthLoginDto, AuthValidateDto } from '../../server/auth/auth.dto';
import {
  EpisodeDto,
  EpisodeScheduleDto,
} from '../../server/episode/episode.dto';
import { PushSubscriptionCreateDto } from '../../server/push_subscription/push_subscription.dto';
import { ShowDto, ShowWithEpisodesDto } from '../../server/show/show.dto';
import {
  SubscriptionCreateDto,
  SubscriptionDto,
  SubscriptionRemoveDto,
  // SubscriptionUpdateDto,
} from '../../server/subscription/subscription.dto';
import { UserDto, UserUpdateDto } from '../../server/user/user.dto';

const axios = Axios.create({
  baseURL: process.env.SERVER_URL,
  withCredentials: true,
});

async function userLogin(loginDto: AuthLoginDto): Promise<void> {
  await axios.post<void>('auth/login', loginDto);
}

async function userValidate(validateDto: AuthValidateDto): Promise<UserDto> {
  const response = await axios.post<UserDto>('auth/validate', validateDto);
  return response.data;
}

async function userGet(): Promise<UserDto> {
  const response = await axios.get<UserDto>('user');
  return response.data;
}

async function userUpdate(updateDto: UserUpdateDto): Promise<UserDto> {
  const response = await axios.put<UserDto>('user', updateDto);
  return response.data;
}

async function showSearch(query: string): Promise<ShowDto[]> {
  const response = await axios.get<ShowDto[]>(`show/search/${query}`);
  return response.data;
}

async function showGet(id: number): Promise<ShowWithEpisodesDto> {
  const response = await axios.get<ShowWithEpisodesDto>(`show/${id}`);
  return response.data;
}

async function episodeGet(id: number): Promise<EpisodeDto> {
  const response = await axios.get<EpisodeDto>(`episode/${id}`);
  return response.data;
}

async function episodeGetByShow(id: number): Promise<EpisodeDto[]> {
  const response = await axios.get<EpisodeDto[]>(`episode/show/${id}`);
  return response.data;
}

async function subscriptionGetAll(): Promise<SubscriptionDto[]> {
  const response = await axios.get<SubscriptionDto[]>('subscription/all');
  return response.data;
}

async function subscriptionCreate(
  createDto: SubscriptionCreateDto
): Promise<SubscriptionDto> {
  const response = await axios.post<SubscriptionDto>('subscription', createDto);
  return response.data;
}

// async function subscriptionUpdate(
//   updateDto: SubscriptionUpdateDto
// ): Promise<SubscriptionDto> {
//   const response = await axios.put<SubscriptionDto>('subscription', updateDto);
//   return response.data;
// }

async function subscriptionRemove(
  removeDto: SubscriptionRemoveDto
): Promise<SubscriptionDto> {
  const response = await axios.delete<SubscriptionDto>('subscription', {
    data: removeDto,
  });
  return response.data;
}

async function episodeScheduleGet(): Promise<EpisodeScheduleDto> {
  const response = await axios.get<EpisodeScheduleDto>('episode/schedule');
  return response.data;
}

async function pushCreate(createDto: PushSubscriptionCreateDto): Promise<void> {
  await axios.post('push_subscription', createDto);
}

export const api = {
  user: {
    login: userLogin,
    validate: userValidate,
    get: userGet,
    update: userUpdate,
  },
  show: {
    search: showSearch,
    get: showGet,
  },
  episode: {
    get: episodeGet,
    getByShow: episodeGetByShow,
    getSchedule: episodeScheduleGet,
  },
  subscription: {
    getAll: subscriptionGetAll,
    create: subscriptionCreate,
    // update: subscriptionUpdate,
    remove: subscriptionRemove,
  },
  push: {
    create: pushCreate,
  },
};
