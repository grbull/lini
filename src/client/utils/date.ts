import { DateTime } from 'luxon';

export function dateFromNow(airstamp: string): string {
  const { locale } = Intl.DateTimeFormat().resolvedOptions();
  return DateTime.fromISO(airstamp, { locale }).toRelative() || airstamp;
}

export function dateToLocaleDate(airstamp: string): string {
  const { locale } = Intl.DateTimeFormat().resolvedOptions();
  return (
    DateTime.fromISO(airstamp, { locale }).toLocaleString(DateTime.DATE_FULL) ||
    airstamp
  );
}

export function dateToLocaleTime(airstamp: string): string {
  const { locale } = Intl.DateTimeFormat().resolvedOptions();
  return (
    DateTime.fromISO(airstamp, { locale }).toLocaleString(
      DateTime.TIME_WITH_SHORT_OFFSET
    ) || ''
  );
}
