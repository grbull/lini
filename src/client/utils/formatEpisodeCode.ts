export function formatEpisodeCode(season: number, episode: number): string {
  let s: string = season.toString(10);
  let e: string = episode.toString(10);
  if (season < 10) {
    s = `0${s}`;
  }
  if (episode < 10) {
    e = `0${e}`;
  }
  return `S${s}E${e}`;
}
