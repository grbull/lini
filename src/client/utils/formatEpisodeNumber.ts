export function formatEpisodeNumber(number: number): string {
  if (number < 10) {
    return `0${number.toString(10)}`;
  }
  return number.toString();
}
