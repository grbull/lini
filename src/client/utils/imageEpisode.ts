// TODO: Create placeholder
const placeHolder = '/placeholder_episode.png';

export function imageEpisode(url: string | null): string {
  if (url) {
    return process.env.IMAGE_PROXY_URL + url;
  }
  return placeHolder;
}
