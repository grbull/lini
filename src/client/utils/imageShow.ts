// TODO: Create placeholder
const placeHolder = '/placeholder_show.png';

export function imageShow(url?: string): string {
  if (url) {
    return process.env.IMAGE_PROXY_URL + url;
  }
  return placeHolder;
}
