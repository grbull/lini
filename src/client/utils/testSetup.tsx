import { configureStore } from '@reduxjs/toolkit';
import { render, RenderResult } from '@testing-library/react';
import { createMemoryHistory, MemoryHistory } from 'history';
import React, { ReactNode } from 'react';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';

import { reducers, RootState } from '../redux/store';

export interface TestSetupProps {
  history?: MemoryHistory;
  state?: Partial<RootState>;
}

export interface TestSetupResult extends RenderResult {
  dispatch: typeof jest.fn;
  history: MemoryHistory<unknown>;
  scrollTo: typeof jest.fn;
}

export function testSetup(
  children: ReactNode,
  props?: TestSetupProps
): TestSetupResult {
  const history = props?.history || createMemoryHistory();
  const store = configureStore({
    reducer: reducers,
    preloadedState: props?.state || undefined,
  });

  // const reduxHistory = [];
  // const dispatch = (action: any): any => reduxHistory.push(action);
  const dispatch = jest.fn();
  store.dispatch = dispatch;

  const scrollTo = jest.fn();
  window.scrollTo = scrollTo;

  const utils = render(
    <Router history={history}>
      <Provider store={store}>{children}</Provider>
    </Router>
  );

  return { dispatch, history, scrollTo, ...utils };
}
