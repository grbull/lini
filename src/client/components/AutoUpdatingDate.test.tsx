/**
 * @jest-environment jsdom
 */

import '@testing-library/jest-dom/extend-expect';

import React from 'react';

import { dateFromNow } from '../utils/date';
import { testSetup } from '../utils/testSetup';
import { AutoUpdatingDate } from './AutoUpdatingDate';

describe('AutoUpdatingDate Component', () => {
  beforeAll(() => {
    // For snapshot consistency we need to fake the date
    Date.now.bind(global.Date);
    global.Date.now = jest.fn(() => 1625584251217);
  });

  it('matches the snapshot', () => {
    const { asFragment } = testSetup(
      <AutoUpdatingDate date="2021-08-05T13:38:35.524Z" format={dateFromNow} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('matches the snapshot without a date', () => {
    const { asFragment } = testSetup(
      <AutoUpdatingDate date={null} format={dateFromNow} />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
