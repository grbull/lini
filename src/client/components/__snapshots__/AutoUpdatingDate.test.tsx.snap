// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoUpdatingDate Component matches the snapshot 1`] = `
<DocumentFragment>
  <time
    datetime="2021-08-05T13:38:35.524Z"
  >
    in 29 days
  </time>
</DocumentFragment>
`;

exports[`AutoUpdatingDate Component matches the snapshot without a date 1`] = `
<DocumentFragment>
  <time>
    N/A
  </time>
</DocumentFragment>
`;
